Hadrons/DistilMatrix.hpp

```diff
@@ -337,6 +560,118 @@ ::fetchDvBatchComponents(unsigned int ibatch, std::vector<unsigned int> time_dil
+::makePinnedPhiComponent(FermionField&              phi_component,
+                   DistillationNoise&               n,
+                   const unsigned int               n_idx,
+                   const unsigned int               D,
+                   const unsigned int               delta_t,
+                   PerambTensor&                    peramb,
+                   LapPack&                         epack)
+{
+    std::array<unsigned int,3> d_coor = n.dilutionCoordinates(D);
+    unsigned int dt = d_coor[Index::t] , dk = d_coor[Index::l] , ds = d_coor[Index::s];
+    
+    //compute at pin_time, insert at t=pin_time
+    const unsigned int pin_time = (dt + delta_t)%nt_;               //TODO: generalise to dilution
+    const unsigned int t        = pin_time;                         //TODO: generalise to dilution 
+
+    std::vector<int> peramb_ts = peramb.MetaData.timeSources;
+    std::vector<int>::iterator itr_dt = std::find(peramb_ts.begin(), peramb_ts.end(), dt);
+    unsigned int idt_peramb = std::distance(peramb_ts.begin(), itr_dt); //gets correspondent index of dt in the tensor obj 
+    const unsigned int nVec = epack.evec.size();
+    const unsigned int Nt_first = g_->LocalStarts()[nd_ - 1];
+    const unsigned int Nt_local = g_->LocalDimensions()[nd_ - 1];
+    if( (pin_time>=Nt_first) and (pin_time<Nt_first+Nt_local) )
+    {
+        tmp3d_ = Zero();
+        for (unsigned int k = 0; k < nVec; k++)
+        {
+            ExtractSliceLocal(evec3d_,epack.evec[k],0,pin_time-Nt_first,nd_ - 1);
+            tmp3d_ += evec3d_ * peramb.tensor(pin_time, k, dk, n_idx, idt_peramb, ds);
+        }
+        InsertSliceLocal(tmp3d_,phi_component,0,t-Nt_first,nd_ - 1);
+    }
+}
+
+template <typename FImpl, typename T, typename Tio>
+void DmfComputation<FImpl,T,Tio>
+::makePinnedRhoComponent(FermionField&          rho_component,
+                   DistillationNoise&           n,
+                   const unsigned int           n_idx,
+                   const unsigned int           D,
+                   const unsigned int           delta_t,
+                   LapPack&                     epack)
+{
+    // abstract this to makePinnedSource() kind of method in DilutedNoise?
+    std::array<unsigned int,3> d_coor = n.dilutionCoordinates(D);
+    unsigned int dt = d_coor[Index::t] , dk = d_coor[Index::l] , ds = d_coor[Index::s];
+
+    typename DistillationNoise::NoiseType   noise(n.getNoise()[n_idx].data(), nt_, epack.eval.size(), Ns);
+
+    //compute at pin_time, insert at t
+    const unsigned int pin_time = (dt + delta_t)%nt_;     //TODO: generalise to dilution
+    const unsigned int t        = pin_time;               //TODO: generalise to dilution
+
+    // adapt to dilution! (add an it loop and untrivialise ik and is loops)
+    
+    const unsigned int Nt_first = g_->LocalStarts()[nd_ - 1];
+    const unsigned int Nt_local = g_->LocalDimensions()[nd_ - 1];
+    if( (pin_time>=Nt_first) and (pin_time<Nt_first+Nt_local) )
+    {
```

Comment:
why the loops over a single component?

Reply:
This is where I will place a loop for general dilution (is commented as TODO). Currently is full dilution hard coded

Reply:
i see, thanks! good to have it set up alreadt